clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -561,6 +584,91 @@ scheduler_tmpl_t<RecordType, ReaderType>::init(
     return STATUS_SUCCESS;
 }
 
+template <typename RecordType, typename ReaderType>
+std::string
+scheduler_tmpl_t<RecordType, ReaderType>::recorded_schedule_component_name(
+    output_ordinal_t output)
+{
+    static const char *const SCHED_CHUNK_PREFIX = "output.";
+    std::ostringstream name;
+    name << SCHED_CHUNK_PREFIX << std::setfill('0') << std::setw(4) << output;
+    return name.str();
+}
+
+template <typename RecordType, typename ReaderType>
+typename scheduler_tmpl_t<RecordType, ReaderType>::scheduler_status_t
+scheduler_tmpl_t<RecordType, ReaderType>::write_recorded_schedule()
+{
+    if (options_.record_ostream == nullptr)
+        return STATUS_ERROR_INVALID_PARAMETER;
+    std::lock_guard<std::mutex> guard(sched_lock_);
+    for (int i = 0; i < static_cast<int>(outputs_.size()); i++) {
+        sched_type_t::stream_status_t status =
+            record_schedule_segment(i, schedule_record_t::FOOTER, 0, 0, 0);
+        if (status != sched_type_t::STATUS_OK)
+            return STATUS_ERROR_FILE_WRITE_FAILED;
+        std::string name = recorded_schedule_component_name(i);
+        std::string err = options_.record_ostream->open_new_component(name);
+        if (!err.empty()) {
+            VPRINT(this, 1, "Failed to open component %s in record file: %s\n",
+                   name.c_str(), err.c_str());
+            return STATUS_ERROR_FILE_WRITE_FAILED;
+        }
+        if (!options_.record_ostream->write(
```

Comment:
No need to close?

Reply:
std::ostream does not have a close method.  raw2trace also takes in ostreams and has no way to close.  It will be closed generally by the caller who owns the lifetime when it goes out of scope there.